File: backend/generators/middle_square.py
"""
Middle-Square Method

Algorithm:
  1. Take the seed and square it.
  2. Zero-pad the result to 2*d digits (where d is the number of digits in the seed).
  3. Extract the middle d digits as the next value.
  4. Repeat.

Parameters:
    seed: initial value (integer with d digits)
    iterations: number of values to generate
"""


def generate(seed: int, iterations: int) -> dict:
    """
    Generate pseudo-random numbers using the Middle-Square method.

    Returns a dict with:
      - sequence: list of generated integers
      - steps: list of dicts with intermediate calculations per iteration
      - normalized: list of floats in [0, 1)
      - digits: number of digits used (d)
    """
    if seed < 0:
        raise ValueError("Seed must be a non-negative integer.")
    if iterations <= 0:
        raise ValueError("Iterations must be a positive integer.")
    if iterations > 10000:
        raise ValueError("Iterations must not exceed 10,000.")

    # Determine number of digits d (use at least 2 digits)
    d = max(len(str(seed)), 2)
    max_val = 10 ** d

    sequence = []
    steps = []
    x = seed

    for i in range(iterations):
        x_prev = x
        squared = x_prev ** 2
        padded = str(squared).zfill(2 * d)
        mid_start = d // 2
        x_next = int(padded[mid_start: mid_start + d])
        normalized = x_next / max_val

        steps.append({
            "iteration": i + 1,
            "x_prev": x_prev,
            "squared": squared,
            "padded": padded,
            "extracted": padded[mid_start: mid_start + d],
            "x_next": x_next,
            "normalized": round(normalized, 6),
        })
        sequence.append(x_next)
        x = x_next

    return {
        "sequence": sequence,
        "steps": steps,
        "normalized": [round(v / max_val, 6) for v in sequence],
        "digits": d,
    }

File: frontend-threejs/src/lib/generators/middle-square.ts
/**
 * Middle-Square Method
 * Algorithm:
 *   1. Square the seed
 *   2. Zero-pad to 2*d digits
 *   3. Extract the middle d digits
 *   4. Repeat
 */

import type { GeneratorResult, MiddleSquareParams, MiddleSquareStepDetail } from './types';

export function generate(params: MiddleSquareParams): GeneratorResult {
  const { seed, iterations, d } = params;

  if (seed < 0) throw new Error('La semilla debe ser un entero no negativo.');
  if (iterations <= 0) throw new Error('Las iteraciones deben ser un entero positivo.');
  if (iterations > 10000) throw new Error('Las iteraciones no deben exceder 10,000.');
  if (d < 2) throw new Error('El número de dígitos d debe ser al menos 2.');
  if (seed >= Math.pow(10, d)) {
    throw new Error(`La semilla no puede tener más de ${d} dígitos.`);
  }

  const maxVal = Math.pow(10, d);
  const midStart = Math.floor(d / 2);

  const sequence: number[] = [];
  const steps: MiddleSquareStepDetail[] = [];
  const seen = new Set<number>();
  let x = seed;

  for (let i = 0; i < iterations; i++) {
    const xPrev = x;
    const squared = xPrev * xPrev;
    const padded = String(squared).padStart(2 * d, '0');
    const leftPart = padded.substring(0, midStart);
    const extracted = padded.substring(midStart, midStart + d);
    const rightPart = padded.substring(midStart + d);
    const xNext = parseInt(extracted, 10);
    const normalized = xNext / maxVal;

    steps.push({
      iteration: i + 1,
      xPrev,
      formula: `${xPrev}² = ${squared} → medio(${padded}) = ${extracted}`,
      xNext,
      normalized: Math.round(normalized * 1000000) / 1000000,
      squared,
      padded,
      extracted,
      leftPart,
      rightPart,
    });

    sequence.push(xNext);

    // Degeneration detection
    if (xNext === 0) {
      return {
        sequence,
        steps,
        normalized: sequence.map((v) => Math.round((v / maxVal) * 1000000) / 1000000),
        digits: d,
        degenerated: { iteration: i + 1, reason: 'zero', value: 0 },
      };
    }
    if (seen.has(xNext)) {
      return {
        sequence,
        steps,
        normalized: sequence.map((v) => Math.round((v / maxVal) * 1000000) / 1000000),
        digits: d,
        degenerated: { iteration: i + 1, reason: 'cycle', value: xNext },
      };
    }
    seen.add(xNext);
    x = xNext;
  }

  return {
    sequence,
    steps,
    normalized: sequence.map((v) => Math.round((v / maxVal) * 1000000) / 1000000),
    digits: d,
    degenerated: null,
  };
}
